### PR TITLE
Added udp support

### PR DESCRIPTION
**Pull request by hatch15**
Added udp support

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -43,6 +48,10 @@ var (
 		},
 	)
 	invalidChars = regexp.MustCompile("[^a-zA-Z0-9_]")
+
+	// Udp
+	bindAddress = flag.String("udp.bind-address", ":8089", "Address on which to listen for udp packets.")
```
Comment (brian-brazil) [inline main.go]: What we've done elsewhere is to default this to the empty string, meaning disabled. In this case it should be safe to default to 9122.

File: main.go [hunk 2]
```diff
@@ -53,10 +62,40 @@ type influxDBSample struct {
...
+				continue
+			}
+
+			bufCopy := make([]byte, n)
+			copy(bufCopy, buf[:n])
+
+			precision := "ns"
+			points, err := models.ParsePointsWithPrecision(bufCopy, time.Now().UTC(), precision)
+			if err != nil {
+				fmt.Printf("error parsing udp packet: %s", err)
```
Comment (brian-brazil) [inline main.go]: We've a logging package you can use for this.

Can you add a metric to count the udp errors?

File: main.go [hunk 3]
```diff
@@ -53,10 +62,40 @@ type influxDBSample struct {
 	Timestamp time.Time
 }
 
+func (c *influxDBCollector) serve() {
```
Comment (brian-brazil) [inline main.go]: serveUdp would be clearer.

**Comment by hatch15** [comment]
Here's an updated patch.

I'm not entirely sure if putting the udpParseErrors counter in the influxDBCollector struct is the right place for it, let me know if i should handle in a different way.

File: main.go [hunk 4]
```diff
@@ -43,6 +48,10 @@ var (
 		},
 	)
 	invalidChars = regexp.MustCompile("[^a-zA-Z0-9_]")
+
+	// Udp
+	bindAddress = flag.String("udp.bind-address", ":9122", "Address on which to listen for udp packets.")
+	listenOnUdp = flag.Bool("udp.listen", false, "Listen on udp socket.")
```
Comment (brian-brazil) [inline main.go]: You can remove this now. I don't really see it happening where having UDP is a problem.

File: main.go [hunk 5]
```diff
@@ -53,16 +62,53 @@ type influxDBSample struct {
...
+}
+
 type influxDBCollector struct {
 	samples map[string]*influxDBSample
 	mu      sync.Mutex
 	ch      chan *influxDBSample
+
+	// Udp
+	conn           *net.UDPConn
+	udpParseErrors prometheus.Counter
```
Comment (brian-brazil) [inline main.go]: Probably best to put it up with the other metric in the top level var.

File: main.go [hunk 6]
```diff
@@ -53,16 +62,53 @@ type influxDBSample struct {
...
 }
 
 func newInfluxDBCollector() *influxDBCollector {
 	c := &influxDBCollector{
 		ch:      make(chan *influxDBSample),
 		samples: map[string]*influxDBSample{},
+		udpParseErrors: prometheus.NewCounter(prometheus.CounterOpts{
+			Namespace: "influxdb",
+			Name:      "exporter_udp_parse_errors",
```
Comment (brian-brazil) [inline main.go]: Counters should end with _total

File: main.go [hunk 7]
```diff
@@ -43,6 +48,10 @@ var (
 		},
 	)
 	invalidChars = regexp.MustCompile("[^a-zA-Z0-9_]")
+
+	// Udp
+	bindAddress = flag.String("udp.bind-address", ":9122", "Address on which to listen for udp packets.")
```
Comment (brian-brazil) [inline main.go]: Keep the flags together.

File: main.go [hunk 8]
```diff
@@ -31,17 +32,28 @@ import (
...
+	bindAddress   = flag.String("udp.bind-address", ":9122", "Address on which to listen for udp packets.")
 	lastPush      = prometheus.NewGauge(
 		prometheus.GaugeOpts{
 			Name: "influxdb_last_push_timestamp_seconds",
 			Help: "Unix timestamp of the last received influxdb metrics push in seconds.",
 		},
 	)
+	udpParseErrors = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Name: "exporter_udp_parse_errors_total",
```
Comment (brian-brazil) [inline main.go]: The prefix should be influxdb to match the above

**Comment by brian-brazil** [comment]
Thanks!

**Pull request merged**